### PR TITLE
[mlir][vector] Add more tests for ConvertVectorToLLVM (2/n)

### PR DESCRIPTION
**Pull request by banach-space**
Adds tests with scalable vectors for the Vector-To-LLVM conversion pass.
Covers the following Ops:
  * vector.outerproduct

### CODE REVIEW

**Review by nujaa — COMMENTED** [review]
LGTM. Please ping when the dependent MRs are merged for approval.
A bit off topic but, This MR got me wondering about the handling of cases where the lhs would be scalable and found out it was considered invalid. e.g. 
```
 // expected-error @+1 {{expected either both or only #2 operand dim to be scalable}}
vector.outerproduct %0, %1 : vector<[4]xf32>, vector<4xf32>
// equivalent to
// for i = 0 to vscale
// | %0i = extract %0[i*4] vector<[4]xf32>-> vector<4xf32>
// | vector.outerproduct %0i, %1 : vector<4xf32>, vector<4xf32>
```
Does it make sense to support this usecase in the future or do we consider this should never be generated and the user never write it ? I have seen another MR recently (from you I think) which did not handle the lowering to a loop. 
```
      // This restriction reflects what's currently supported in terms of
      // scalable vectors. However, we could relax this if there's a use case.
```
In case I get into sorting this out, do you think we should have some separate scalable vectors canonicalization patterns a bit like ArmSMEToSCF but without the SME dependency or simply have lowerVectorToLLVM support this case is enough ?

**Comment by banach-space** [comment]
> Does it make sense to support this usecase in the future or do we consider this should never be generated and the user never write it ?

It's always good to have a motivating example - can you think of a situation where that could be used? Nothing on my mind today, but I tend to look at a rather limited set of examples.

In general, I always advocate for good balance between complexity and what's actually needed. In particular with scalable vectors, we tend to discover that we don't really need every single pattern in the Vector dialect to support them. And we still manage to generate good code 😅  (just as an example)

Having said that, I'm not opposed to improving the support for scalable vectors.

> I have seen another MR recently (from you I think) which did not handle the lowering to a loop.
> 
> ```
>       // This restriction reflects what's currently supported in terms of
>       // scalable vectors. However, we could relax this if there's a use case.
> ```
> 
> In case I get into sorting this out, do you think we should have some separate scalable vectors canonicalization patterns a bit like ArmSMEToSCF but without the SME dependency or simply have lowerVectorToLLVM support this case is enough ?

Depends on complexity. For more involved logic, I'd start with something "hidden" in the SVE dialect. And then, if we find that particularly useful, we could "elevate" that to Vector. Having said that, this particular functionality would be valid for any implementation of scalable vectors, not only SVE.

Btw, I really appreciate all the reviews and wanted to say that I'm always happy to return the favour. In particular, I'm reviewing all Vector tests and if you ever feel like taking a look at one of the tests yourself, here's a list that I use to co-ordinate this:
* https://github.com/orgs/llvm/projects/23

Sharing mostly so that you are aware :)

**Comment by nujaa** [comment]
Alright, thanks for clarification.
One usecase I find is, if scalable vectors(resp. tiles) have been generated but the machine does not support scalable vectors (resp 2d scalable). Let's consider there are no sve/sme specific ops but e.g. `vector.outerproduct <[4]xf32> <[4]xf32>` . Should lower-to-llvm be able to lower scalable ops to fixed size or compiling without sve/sme/...  flags sets vscale to 1 and shall never build e.g. z registers ?
Mmmh, In any case, if vscale is not supported, there is no loop to be generated. So only remains the case we have 2d scalable vectors but only have SVE. That's quite niche.





> Btw, I really appreciate all the reviews and wanted to say that I'm always happy to return the favour. In particular, I'm reviewing all Vector tests and if you ever feel like taking a look at one of the tests yourself, here's a list that I use to co-ordinate this:
> 
> * https://github.com/orgs/llvm/projects/23
> 
> Sharing mostly so that you are aware :)

Thanks for your kind words. I have to deliver something downstream at present. I have a few features in mind for the short term future. I'll bookmark your status 😉

**Comment by banach-space** [comment]
> LGTM. Please ping when the dependent MRs are merged for approval.

Ping :)

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir [hunk 1]
```diff
@@ -663,6 +734,16 @@ func.func @masked_float_add_outerprod(%arg0: vector<2xf32>, %arg1: f32, %arg2: v
...
 // CHECK:           %[[VAL_9:.*]] = arith.select %[[VAL_3]], %[[VAL_8]], %[[VAL_2]] : vector<2xi1>, vector<2xf32>
 
+func.func @masked_float_add_outerprod_scalable(%arg0: vector<[2]xf32>, %arg1: f32, %arg2: vector<[2]xf32>, %m: vector<[2]xi1>) -> vector<[2]xf32> {
+  %0 = vector.mask %m { vector.outerproduct %arg0, %arg1, %arg2 {kind = #vector.kind<add>} : vector<[2]xf32>, f32 } : vector<[2]xi1> -> vector<[2]xf32>
+  return %0 : vector<[2]xf32>
+}
+
+// CHECK-LABEL:   func.func @masked_float_add_outerprod_scalable
+// CHECK-SAME:                                          %[[VAL_0:.*]]: vector<[2]xf32>, %[[VAL_1:.*]]: f32, %[[VAL_2:.*]]: vector<[2]xf32>, %[[VAL_3:.*]]: vector<[2]xi1>) -> vector<[2]xf32> {
+// CHECK:           %[[VAL_8:.*]] = llvm.intr.fmuladd(%[[VAL_0]], %{{.*}}, %[[VAL_2]])  : (vector<[2]xf32>, vector<[2]xf32>, vector<[2]xf32>) -> vector<[2]xf32>
```
Comment (nujaa) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:744]: NIT:missing VAL_[4-7]

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir [hunk 2]
```diff
@@ -663,6 +734,16 @@ func.func @masked_float_add_outerprod(%arg0: vector<2xf32>, %arg1: f32, %arg2: v
 // CHECK:           %[[VAL_8:.*]] = llvm.intr.fmuladd(%[[VAL_0]], %{{.*}}, %[[VAL_2]])  : (vector<2xf32>, vector<2xf32>, vector<2xf32>) -> vector<2xf32>
 // CHECK:           %[[VAL_9:.*]] = arith.select %[[VAL_3]], %[[VAL_8]], %[[VAL_2]] : vector<2xi1>, vector<2xf32>
 
+func.func @masked_float_add_outerprod_scalable(%arg0: vector<[2]xf32>, %arg1: f32, %arg2: vector<[2]xf32>, %m: vector<[2]xi1>) -> vector<[2]xf32> {
```
Comment (nujaa) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:737]: Is renaming `arg[0-9]*` to `vec_[0-9]*`, ... in your future plans ?

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir [hunk 3]
```diff
@@ -663,6 +734,16 @@ func.func @masked_float_add_outerprod(%arg0: vector<2xf32>, %arg1: f32, %arg2: v
...
+  %0 = vector.mask %m { vector.outerproduct %arg0, %arg1, %arg2 {kind = #vector.kind<add>} : vector<[2]xf32>, f32 } : vector<[2]xi1> -> vector<[2]xf32>
+  return %0 : vector<[2]xf32>
+}
+
+// CHECK-LABEL:   func.func @masked_float_add_outerprod_scalable
+// CHECK-SAME:                                          %[[VAL_0:.*]]: vector<[2]xf32>, %[[VAL_1:.*]]: f32, %[[VAL_2:.*]]: vector<[2]xf32>, %[[VAL_3:.*]]: vector<[2]xi1>) -> vector<[2]xf32> {
```
Comment (nujaa) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:743]: Shall we align arg definition with function body as in the other tests ?

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir (hunk 2 above)
Comment (banach-space) [reply mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:737]: This file consistently uses `%arg0` and `%arg1`, so renaming would be just extra noise. In the other files, we were mixing different styles, so I've renamed everything (to `%vec`) for consistency. But, IMHO, the actual name isn't as important as consistency 😅

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir (hunk 1 above)
Comment (banach-space) [reply mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:744]: That's intentional and consistent with other tests :) I am only testing the key stuff (to reduce noise).

**Review by banach-space — COMMENTED** [review]
Thanks for the review, sending update shortly!

**Comment by banach-space** [comment]
@nujaa I have noticed that you "click"ed merge without _accepting_  ;-) Could you formally "+1" as well? 

Thanks  for the review and all the support with this 🙏🏻 !

**Review by nujaa — COMMENTED** [review]
LGTM

**Comment by banach-space** [comment]
> I thought It would not let me merge without Approval so I imagined it was approved 🤒

It's also OK to leave for the author to merge it ;-) In fact, I might not get any buildbot notifications otherwise 🤔 (perhaps that has changed)

**Pull request merged**